### DISCUSSION TITLE
How to exclude seeds from loading into assets when using @dbt_assets?

### DISCUSSION THREAD

**Discussion by mmutso-boku** [dagster-io/dagster#16010] [Q&A] [answered]
Hello,

Currently working on migrating from 1.3.14 to 1.4.x, and ran into a bit of a roadblock regarding dbt assets.
Since a lot of parameters got deprecated in `load_assets_from_dbt_project` and the suggested replacement being `@dbt_assets`, I tried to convert all of my dbt assets to use this new decorator.

The thing I failed to achieve was excluding seeds from being loaded as assets. From digging into the source code and experimenting with the `exclude` parameter, it seems to me that something similar to `use_build_command=False` in `load_assets_from_dbt_project` is missing from `@dbt_assets` OR there is a bug with the `exclude` parameter not extending to seeds (not 100% sure in this though).

Have you ran into this problem also, and have you manage to solve it somehow?

Thanks in advance.

**Comment by rexledesma** [comment]
You can add `exclude="resource_type:seed"`. For example, in jaffle shop, you'll see the following:

```python
@dbt_assets(
    manifest=dbt_manifest_path,
    exclude="resource_type:seed",
)
def jaffle_shop_dbt_assets(context: OpExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(["run"], context=context)
```

![127 0 0 1_3000_locations_jaffle_dagster_asset-groups_default (1)](https://github.com/dagster-io/dagster/assets/16431325/6b86da8c-ecac-4952-b586-f07ef09cd765)

Let me know if you were expecting different behavior.

**Comment by mmutso-boku** [comment]
@rexledesma The `exclude="resource_type:seed"` seems to have no effect in my case.

I guess the specific scenario is - excluding seeds even if they are connected with/upstream of the selected model.

![image](https://github.com/dagster-io/dagster/assets/116641288/6a636303-3067-4a6e-8706-ceb81d287c81)

```
@dbt_assets(manifest=MANIFEST_PATH,
            exclude='resource_type:seed',
            select='the_dbt_asset',
            partitions_def=DailyPartitionsDefinition(datetime(2023, 5, 14)),
            dagster_dbt_translator=CustomDbtTranslator())
def the_dbt_asset(context: AssetExecutionContext, dbt: DbtCliResource):
    start, end = context.partition_key_range
    dbt_vars = {"start": str(start), "end": str(end)}
    dbt_args = ["run", "--vars", json.dumps(dbt_vars)]

    yield from dbt.cli(dbt_args, context=context)
```

**Reply by rexledesma** [reply, marked as answer]
One solution that I thought of is that you could modify the manifest in place to remove references of seeds. That way, when Dagster introspects the manifest to generate software-defined assets, it will not generate a reference for the dbt seed. Will that solve your issue?

```python
# assets.py
from dagster import OpExecutionContext
from dagster_dbt import DbtCliResource, dbt_assets

from .constants import dbt_manifest_path


@dbt_assets(manifest=dbt_manifest_path)
def jaffle_shop_dbt_assets(context: OpExecutionContext, dbt: DbtCliResource):
    dbt.cli(["seed"], context=context).wait()
    yield from dbt.cli(["run"], context=context).stream()

```

```python
# constants.py
import json
import os
from pathlib import Path

from dagster_dbt import DbtCliResource

dbt_project_dir = Path(__file__).joinpath("..", "..", "..", "jaffle_shop").resolve()
dbt = DbtCliResource(project_dir=os.fspath(dbt_project_dir))

# If DAGSTER_DBT_PARSE_PROJECT_ON_LOAD is set, a manifest will be created at runtime.
# Otherwise, we expect a manifest to be present in the project's target directory.
if os.getenv("DAGSTER_DBT_PARSE_PROJECT_ON_LOAD"):
    dbt_parse_invocation = dbt.cli(["parse"], manifest={}).wait()
    dbt_manifest_path = dbt_parse_invocation.target_path.joinpath("manifest.json")
else:
    dbt_manifest_path = dbt_project_dir.joinpath("target", "manifest.json")


def remove_seeds_from_manifest(dbt_manifest_path: Path):
    # Remove seeds
    manifest = json.loads(dbt_manifest_path.read_bytes())
    manifest_nodes = manifest["nodes"]

    # Remove seeds from dependencies
    for unique_id, dbt_resource_props in manifest_nodes.items():
        dbt_resource_props["depends_on"]["nodes"] = [
            node
            for node in dbt_resource_props["depends_on"].get("nodes", [])
            if not node.startswith("seed")
        ]

    # Remove seeds from nodes
    manifest["nodes"] = {
        unique_id: dbt_resource_props
        for unique_id, dbt_resource_props in manifest_nodes.items()
        if not unique_id.startswith("seed")
    }

    # Remove seeds from parent and child maps
    for mapping in ["child_map", "parent_map"]:
        manifest[mapping] = {
            unique_id: [node for node in child_nodes if not node.startswith("seed")]
            for unique_id, child_nodes in manifest[mapping].items()
            if not unique_id.startswith("seed")
        }

    dbt_manifest_path.write_text(json.dumps(manifest))
    print(dbt_manifest_path)


remove_seeds_from_manifest(dbt_manifest_path)

```

![127 0 0 1_3000_locations_jaffle_dagster_asset-groups_default_lineage_](https://github.com/dagster-io/dagster/assets/16431325/b1df0589-7120-4410-91b0-653be67a152f)
